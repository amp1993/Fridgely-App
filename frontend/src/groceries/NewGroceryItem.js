import React, { useEffect } from "react";
import { Formik, Form, Field, ErrorMessage } from 'formik';
import { Navigate, Redirect, useNavigate } from 'react-router-dom';
import Select from 'react-select';
import { unitOfMeasureOptions, categoryOptions } from "../common/itemFormOptions"
import FridgelyApi from "../api/api"
import "./NewGroceryItem.css"
import { Card, CardBody, CardTitle } from "reactstrap";

const NewGroceryItem = ({ user, updateShowPopup, showPopup, userLoggedIn }) => {
    const navigate = useNavigate();



    const initialValues = {
        productName: '',
        categoryName: '',
        unitOfMeasure: '',
        quantityInGroceryList: '',
        username: user
    };

    const fieldsToValidate = ['productName', 'categoryName', 'quantityInGroceryList'];

    const handleSubmit = async (values, { setSubmitting }) => {
        try {

            const productName = values.productName;
            //Check if item already exists
            const existingItem = await FridgelyApi.findGroceryItem(productName, user);
            if (existingItem === productName) {
                //If item exists, have message popUp = true
                updateShowPopup(true);
                // Hide the pop-up after 3 seconds
                setTimeout(() => {
                    updateShowPopup(false);

                }, 3000);
            } else {
                //If item does not exist, proceed with creating item
                await FridgelyApi.createGroceryItem(values, user);
                setSubmitting(false);
                navigate('/groceries');

            }

        } catch (error) {
            console.error("Error creating grocery item:", error);
            setSubmitting(false);
        }
    };

    return (
        <>
            {showPopup && (
                <div className="popup">
                    <p>Item already exists in the grocery list!</p>
                </div>
            )}

            {user ? (

                <div className="form-container">
                    <Card>
                        <CardTitle>
                            <h1 className="login-title">Add New Item to Grocery</h1>

                        </CardTitle>
                        <CardBody>
                            <Formik
                                initialValues={initialValues}
                                validate={values => {
                                    const errors = {};
                                    fieldsToValidate.forEach(field => {
                                        if (!values[field]) {
                                            errors[field] = 'Required';
                                        }
                                    });
                                    return errors;
                                }}
                                onSubmit={handleSubmit}
                            >
                                {({ values, setFieldValue, isSubmitting }) => (
                                    <Form>
                                        <label htmlFor="productName" className="label">Product Name</label>
                                        <Field type="text" name="productName" className="field" />
                                        <ErrorMessage name="productName" component="div" className="error-message" />

                                        <label htmlFor="categoryName" className="label">Category</label>
                                        <Select
                                            name="categoryName"
                                            options={categoryOptions}
                                            className="field"
                                            value={categoryOptions.find(option => option.value === values.categoryName)}
                                            onChange={option => setFieldValue('categoryName', option.value)}
                                            isSearchable
                                        />
                                        <ErrorMessage name="categoryName" component="div" className="error-message" />

                                        <label htmlFor="unitOfMeasure" className="label">Unit of Measure</label>
                                        <Select
                                            name="unitOfMeasure"
                                            options={unitOfMeasureOptions}
                                            className="field"
                                            value={unitOfMeasureOptions.find(option => option.value === values.unitOfMeasure)}
                                            onChange={option => setFieldValue('unitOfMeasure', option.value)}
                                            isSearchable
                                        />
                                        <ErrorMessage name="unitOfMeasure" component="div" className="error-message" />

                                        <label htmlFor="quantityInGroceryList" className="label">Quantity</label>
                                        <Field type="number" name="quantityInGroceryList" className="field" />
                                        <ErrorMessage name="quantityInGroceryList" component="div" className="error-message" />

                                        <div className="button-container">
                                            <button type="submit" disabled={isSubmitting}>Submit</button>
                                        </div>
                                    </Form>
                                )}
                            </Formik>
                        </CardBody>
                    </Card>

                </div>) : (<Navigate to="/" />)}


        </>
    );
};

export default NewGroceryItem;
