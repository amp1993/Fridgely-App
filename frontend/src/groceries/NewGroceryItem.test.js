import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import NewGroceryItem from './NewGroceryItem';

it('renders without crashing', function() {
  render(
    <BrowserRouter>
      <NewGroceryItem user="testuser" updateShowPopup={() => {}} showPopup={false} userLoggedIn={true} />
    </BrowserRouter>
  );
});

it('matches snapshot', function() {
  const { asFragment } = render(
    <BrowserRouter>
      <NewGroceryItem user="testuser" updateShowPopup={() => {}} showPopup={false} userLoggedIn={true} />
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it('displays form fields correctly', function() {
  const { container } = render(
    <BrowserRouter>
      <NewGroceryItem user="testuser" updateShowPopup={() => {}} showPopup={false} userLoggedIn={true} />
    </BrowserRouter>
  );

  expect(container.querySelector('input[name="productName"]')).toBeInTheDocument();
  expect(container.querySelector('input[name="categoryName"]')).toBeInTheDocument();
  expect(container.querySelector('input[name="unitOfMeasure"]')).toBeInTheDocument();
  expect(container.querySelector('input[name="quantityInGroceryList"]')).toBeInTheDocument();
});

