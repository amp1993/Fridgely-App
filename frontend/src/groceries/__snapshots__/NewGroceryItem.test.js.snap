// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="form-container"
  >
    <div
      class="card"
    >
      <div
        class="card-title"
      >
        <h1
          class="login-title"
        >
          Add New Item to Grocery
        </h1>
      </div>
      <div
        class="card-body"
      >
        <form
          action="#"
        >
          <label
            class="label"
            for="productName"
          >
            Product Name
          </label>
          <input
            class="field"
            name="productName"
            type="text"
            value=""
          />
          <label
            class="label"
            for="categoryName"
          >
            Category
          </label>
          <div
            class="field css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-4-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class=" css-13cymwt-control"
            >
              <div
                class=" css-1fdsijx-ValueContainer"
              >
                <div
                  class=" css-1jqq78o-placeholder"
                  id="react-select-4-placeholder"
                >
                  Select...
                </div>
                <div
                  class=" css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-4-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="react-select-4-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="categoryName"
              type="hidden"
              value=""
            />
          </div>
          <label
            class="label"
            for="unitOfMeasure"
          >
            Unit of Measure
          </label>
          <div
            class="field css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-5-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class=" css-13cymwt-control"
            >
              <div
                class=" css-1fdsijx-ValueContainer"
              >
                <div
                  class=" css-1jqq78o-placeholder"
                  id="react-select-5-placeholder"
                >
                  Select...
                </div>
                <div
                  class=" css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-5-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="react-select-5-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="unitOfMeasure"
              type="hidden"
              value=""
            />
          </div>
          <label
            class="label"
            for="quantityInGroceryList"
          >
            Quantity
          </label>
          <input
            class="field"
            name="quantityInGroceryList"
            type="number"
            value=""
          />
          <div
            class="button-container"
          >
            <button
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
